src/category_theory/sites/grothendieck.lean

```diff
@@ -0,0 +1,317 @@
+2. If `S ∈ J X` then its pullback along any `h : Y ⟶ X` is in `J Y`.
+3. If `S ∈ J X` and `R` is a sieve on `X`, then provided that the pullback of `R` along any arrow
+   `f : Y ⟶ X` in `S` is in `J Y`, we have that `R` itself is in `J X`.
+
+A sieve `S` on `X` is referred to as `J`-covering, (or just covering), if `S ∈ J X`.
+-/
+@[ext]
+structure grothendieck_topology :=
+(sieves : Π (X : C), set (sieve X))
+(top_mem' : ∀ X, ⊤ ∈ sieves X)
+(pullback_stable' : ∀ ⦃X Y : C⦄ ⦃S : sieve X⦄ (f : Y ⟶ X), S ∈ sieves X → S.pullback f ∈ sieves Y)
+(transitive' : ∀ ⦃X⦄ ⦃S : sieve X⦄ (hS : S ∈ sieves X) (R : sieve X),
+              (∀ ⦃Y⦄ ⦃f : Y ⟶ X⦄, S.arrows f → R.pullback f ∈ sieves Y) → R ∈ sieves X)
+
+namespace grothendieck_topology
+
+instance : has_coe_to_fun (grothendieck_topology C) :=
+⟨_, λ J, J.sieves⟩
+
+variables {C} {X Y : C} {S R : sieve X}
+variables (J : grothendieck_topology C)
+
+@[simp] lemma mem_sieves_iff_coe : S ∈ J.sieves X ↔ S ∈ J X := iff.rfl
+
+-- Also known as the maximality axiom.
+@[simp] lemma top_mem (X : C) : ⊤ ∈ J X := J.top_mem' X
+-- Also known as the stability axiom.
+@[simp] lemma pullback_stable (f : Y ⟶ X) (hS : S ∈ J X) : S.pullback f ∈ J Y :=
+J.pullback_stable' f hS
+lemma transitive (hS : S ∈ J X) (R : sieve X)
+  (h : ∀ ⦃Y⦄ ⦃f : Y ⟶ X⦄, S.arrows f → R.pullback f ∈ J Y) :
+  R ∈ J X :=
+J.transitive' hS R h
+
+lemma covering_of_eq_top : S = ⊤ → S ∈ J X := λ h, h.symm ▸ J.top_mem X
+
+lemma superset_covering (Hss : S ≤ R) (sjx : S ∈ J X) : R ∈ J X :=
+begin
+  apply J.transitive sjx R (λ Y f hf, _),
+  apply covering_of_eq_top,
+  rw [← top_le_iff, ← S.pullback_eq_top_of_mem hf],
+  apply sieve.pullback_monotone _ Hss,
+end
+
+lemma intersection_covering (rj : R ∈ J X) (sj : S ∈ J X) : R ⊓ S ∈ J X :=
+begin
+  apply J.transitive rj _ (λ Y f Hf, _),
+  rw [sieve.pullback_inter, R.pullback_eq_top_of_mem Hf],
+  simp [sj],
+end
+
+@[simp]
+lemma intersection_covering_iff : R ⊓ S ∈ J X ↔ R ∈ J X ∧ S ∈ J X :=
+⟨λ h, ⟨J.superset_covering inf_le_left h, J.superset_covering inf_le_right h⟩,
+ λ t, intersection_covering _ t.1 t.2⟩
+
+/--
+The sieve `S` on `X` `J`-covers an arrow `f` to `X` if `S.pullback f ∈ J Y`.
+This definition is an alternate way of presenting a Grothendieck topology.
+-/
+def covers (S : sieve X) (f : Y ⟶ X) : Prop := S.pullback f ∈ J Y
+
+lemma covers_iff (S : sieve X) (f : Y ⟶ X) : J.covers S f ↔ S.pullback f ∈ J Y :=
+iff.rfl
+
+lemma covering_iff_covers_id (S : sieve X) : S ∈ J X ↔ J.covers S (𝟙 X) :=
+by simp [covers_iff]
+
+/-- The maximality axiom in 'arrow' form: Any arrow `f` in `S` is covered by `S`. -/
+lemma arrow_max (f : Y ⟶ X) (S : sieve X) (hf : S.arrows f) : J.covers S f :=
+begin
+  rw [covers, (sieve.pullback_eq_top_iff_mem f).1 hf],
+  apply J.top_mem,
+end
+
+/-- The stability axiom in 'arrow' form: If `S` covers `f` then `S` covers `g ≫ f` for any `g`. -/
+lemma arrow_stab (f : Y ⟶ X) (S : sieve X) (h : J.covers S f) {Z : C} (g : Z ⟶ Y) :
```

Comment:
Could we uniformly abbreviate `stable`? My preference would be by not abbreviating it at all.